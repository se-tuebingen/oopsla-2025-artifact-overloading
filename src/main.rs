use overloading::compile_expr;
use overloading::syntax::parse_expr;

use clap::Parser;

#[derive(Parser)]
#[command(name = "overloading")]
struct Args {
    /// Input file to process
    filename: String,

    /// Use short output format
    #[arg(short, long)]
    short: bool,
}

/// Main entry point for the native CLI.
fn main() {
    let args = Args::parse();

    let content = std::fs::read_to_string(&args.filename).expect("Failed to read file");
    match parse_expr(&content) {
        Ok(expr) => {
            println!("{}", compile_expr(&expr, args.short));
        }
        Err(err) => {
            panic!("Failed to parse expression: {}", err);
        }
    }
}

/// Snapshot tests for the compiler examples.
#[cfg(test)]
#[test]
fn test_compiler_examples() {
    use insta::{assert_json_snapshot, glob, with_settings};

    glob!("../", "examples/*.over", |path| {
        let input = std::fs::read_to_string(path).unwrap();
        let output = match parse_expr(&input) {
            Ok(expr) => overloading::compile_expr_json(&expr, true),
            Err(err) => overloading::json::JsonProcessingResult::InferenceError {
                expression: input,
                error: err,
            },
        };

        // Use the filename (without extension) as snapshot suffix
        let filename = path.file_stem().unwrap().to_str().unwrap();
        with_settings!({
            snapshot_path => "../examples/snapshots",
            snapshot_suffix => filename,
            sort_maps => true,
        }, {
            assert_json_snapshot!(output);
        });
    });
}
