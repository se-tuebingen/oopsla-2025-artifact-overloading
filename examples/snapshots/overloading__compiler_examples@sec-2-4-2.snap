---
source: src/main.rs
expression: output
input_file: examples/sec-2-4-2.over
---
{
  "status": "success",
  "expression": "let defaultNum = a‹0, 0›\nb‹defaultNum, \"\"›",
  "dimension_spans": "[spans]",
  "inferred_type": "over_b_1",
  "worlds": {
    "type": "many",
    "count": 4.0
  },
  "solutions": [
    [
      "a₁",
      "b₁"
    ],
    [
      "a₂",
      "b₁"
    ],
    [
      "a₁",
      "b₂"
    ],
    [
      "a₂",
      "b₂"
    ]
  ],
  "constraints": [
    {
      "from": "Int",
      "choices": [
        "a₁"
      ],
      "to": "over_a_0"
    },
    {
      "from": "Double",
      "choices": [
        "a₂"
      ],
      "to": "over_a_0"
    },
    {
      "from": "over_a_0",
      "choices": [
        "b₁"
      ],
      "to": "over_b_1"
    },
    {
      "from": "String",
      "choices": [
        "b₂"
      ],
      "to": "over_b_1"
    }
  ],
  "bounds": [
    {
      "variable": "over_a_0",
      "lower": [
        {
          "type_name": "Double",
          "choices": [
            "a₂"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "a₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "over_b_1",
          "choices": [
            "b₁"
          ]
        }
      ]
    },
    {
      "variable": "over_b_1",
      "lower": [
        {
          "type_name": "Double",
          "choices": [
            "a₂",
            "b₁"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "a₁",
            "b₁"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "b₂"
          ]
        }
      ],
      "upper": []
    }
  ],
  "errors": []
}
